### PR TITLE
CI: add ML ROCm stack

### PR DESCRIPTION
**Pull request by afzpatel**
The ML ROCm stack was removed here: https://github.com/spack/spack/pull/43825. Now that py-tensorflow+rocm is supported with spack built rocm, we can add this back.

### CODE REVIEW

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      require:
+      - target=x86_64_v3
+      - ~cuda
+      - +rocm
+      - amdgpu_target=gfx90a
+    gl:
+      require: "osmesa"
+    py-torch:
+      # Does not yet support Spack-installed ROCm
+      require: ~rocm
```
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: Let's remove this so we can actually test PyTorch too

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    - py-tensorflow-probability
+
+    # XGBoost
+    - py-xgboost
+
+  ci:
+    pipeline-gen:
+    - build-job:
+        image:
+          name: ghcr.io/spack/linux-ubuntu22.04-x86_64_v2:v2024-01-29
```
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]:
```suggestion
          name: ghcr.io/spack/ubuntu-22.04:v2024-05-07
```
I created a newer image with a newer version of GCC

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    # Hugging Face
+    - py-transformers
+
+    # JAX
+    - py-jax
+    - py-jaxlib
+
+    # Keras
+    # - py-keras backend=tensorflow
```
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: Would like to test this too

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 4]
```diff
@@ -726,9 +726,32 @@ ml-linux-x86_64-cuda-build:
...
+# Machine Learning - Linux x86_64 (ROCm)
+########################################
+.ml-linux-x86_64-rocm:
+  extends: [ ".linux_x86_64_v3" ]
+  variables:
+    SPACK_CI_STACK_NAME: ml-linux-x86_64-rocm
+
+ml-linux-x86_64-rocm-generate:
+  extends: [ ".generate-x86_64", .ml-linux-x86_64-rocm, ".tags-x86_64_v4" ]
+  image: ghcr.io/spack/linux-ubuntu22.04-x86_64_v2:v2024-01-29
```
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]:
```suggestion
  image: ghcr.io/spack/ubuntu-22.04:v2024-05-07
```

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 5]
```diff
@@ -726,9 +726,32 @@ ml-linux-x86_64-cuda-build:
...
+        job: ml-linux-x86_64-rocm-generate
+    strategy: depend
+  needs:
+    - artifacts: True
+      job: ml-linux-x86_64-rocm-generate
+
+########################################
 # Machine Learning - Darwin aarch64 (MPS)
-#########################################
+########################################
```
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Can you undo this change to the #?

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Comment by adamjstewart** [comment]
I think we still need to add a ROCm variant to JAX too

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml (hunk 1 above)
Comment (afzpatel) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: py-torch is still not supported for Spack-installed ROCm

**Comment by adamjstewart** [comment]
CI is failing due to the ROCm version naming convention. Keras needs TF 2.16.1:2.16.999, but you have 2.16 with no minor version information.

**Comment by afzpatel** [comment]
> CI is failing due to the ROCm version naming convention. Keras needs TF 2.16.1:2.16.999, but you have 2.16 with no minor version information.

There's no minor version information for the release we're using: https://github.com/ROCm/tensorflow-upstream/releases/tag/r2.16-rocm-enhanced. We could rename the version in the spack recipe to 2.16.1-rocm-enhanced. 

We could also change py-keras to: `depends_on("py-tensorflow@2.16", when="@3.0: backend=tensorflow")`, since the py-tensorflow recipe only has versions 2.16.1 and 2.16.2.

**Comment by adamjstewart** [comment]
> We could also change py-keras

I would like to avoid modifying other packages just because the ROCm fork does not follow the same versioning scheme as the upstream repo. Note that the master version of Keras now requires `~=2.16.2`, so this solution wouldn't work anyway.

> There's no minor version information

What minor version is `r2.16-rocm-enhanced` most similar to? Was it branched off of upstream when 2.16.0 was created? Do you backport the same commits as upstream? Or is the development basically independent of upstream?

**Comment by hsharsha** [comment]
> What minor version is r2.16-rocm-enhanced most similar to? Was it branched off of upstream when 2.16.0 was created? Do you backport the same commits as upstream? Or is the development basically independent of upstream?

It is 2.16.1.  We backport relevant commits from upstream to this branch and tag the release by bumping minor version. As it was done for the first time, I skipped minor version tag.

**Comment by adamjstewart** [comment]
Great! Maybe it would be possible to add a new tag with the patch version and use that in the name. Either way, I think we can rename the version in the Spack package now that we know what patch version it is.

**Comment by afzpatel** [comment]
py-tensorflow version 2.16-rocm-enhanced has been renamed to 2.16.1-rocm-enhanced(https://github.com/spack/spack/commit/4447d3339c39753489cdc14c5bab6210107e2d65)

**Comment by adamjstewart** [comment]
Awesome, now we just need to figure out how to get `py-tensorflow+rocm` to build.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml (hunk 1 above)
Comment (adamjstewart) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: In that case, let's remove this and then comment out `py-torch` below so we aren't building yet another CPU copy.

Comment (adamjstewart) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: We could also add externally installed ROCm to the CI image if that would help with testing.

Comment (afzpatel) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: I've commented out py-torch and related packages.

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    - py-transformers
+
+    # JAX
+    - py-jax
+    - py-jaxlib
+
+    # Keras
+    - py-keras backend=tensorflow
+    - py-keras backend=jax
+    - py-keras backend=torch
```
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: Will need to comment out torch here as well

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+      - +rocm
+      - amdgpu_target=gfx90a
+    gl:
+      require: "osmesa"
+    mpi:
+      require: openmpi
+
+  specs:
+    # Horovod
+    - py-horovod
```
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: Might be building with torch, you can tell it to use TF

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+    # Hugging Face
+    - py-transformers
```
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml:20]: Torch

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml (hunk 7 above)
Comment (afzpatel) [reply share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: I commented it out since it requires py-tensorflow-estimator if it uses TF. That wouldn't work since py-tensorflow-estimator doesn't have a 2.16 version

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  specs:
+    # Horovod
+    # - py-horovod
+
+    # Hugging Face
+    - py-transformers
+
+    # JAX
+    - py-jax
+    - py-jaxlib
```
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: Let's comment out JAX until someone adds a ROCm variant

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+    # Keras
+    - py-keras backend=tensorflow
+    - py-keras backend=jax
```
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: Same here

File: share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    - py-tensorflow
+    - py-tensorflow-datasets
+    # version 2.16 is not available
+    # - py-tensorflow-estimator
+    - py-tensorflow-hub
+    - py-tensorflow-metadata
+    - py-tensorflow-probability
+
+    # XGBoost
+    - py-xgboost
```
Comment (adamjstewart) [inline share/spack/gitlab/cloud_pipelines/stacks/ml-linux-x86_64-rocm/spack.yaml]: Let's comment this out until someone adds ROCm support

**Review by adamjstewart — APPROVED** [review]
This looks great, glad we finally have this tested in CI! Looking forward to further progress on ROCm support across other popular ML packages and upstreaming some of these forks/patches to the original developers!
